texastribune/files
# Integrate components from elements package

#### What's this PR do?

Copies custom elements from the [elements repo](https://github.com/texastribune/Elements) into this repo and removes the elements repo as a remote dependency.

#### Why are we doing this? How does it help us?

This repo has previously relied on importing web components that were stored in a separate in-house components library called [elements](https://github.com/texastribune/Elements). Although I wasn't working at the Tribune at the time, I understand the decision to import these components as a node module was made so that they could be readily reused for other Tribune projects that might use web components.

Unfortunately, since both this package and the elements package were created around 2018/2019 we have not used the elements package in any other Tribune projects and have no plans to do so. In general, the momentum and enthusiasm surrounding web components appears to have [faded somewhat](https://blog.logrocket.com/what-happened-to-web-components/#:~:text=It's%20not%20due%20to%20a,of%20entry%20is%20too%20high.) in recent years. 

After discussion with the Tribune engineering team it was decided that copying the existing elements into this repo would make local development with the files app and the tribune project much easier. It also means that we can better customize these elements to the precise needs of the files app without worrying about whether the components in the elements package are overly coupled to the files app.

An alternative solution to these issues would be to simply extend some of the elements from the elements package as needed to create new custom elements that have new or modified behavior. This was somewhat tempting but again, since we aren't using the elements package in any other project, it was judged that this would needlessly add another layer of complexity to the project. Also, owing to the way CSS is encapsulated in web components, simply extending the components in the elements package does not fully circumvent the difficulty in modifying their CSS.

By integrating these components directly into this repo we can simplify the structure of the files app and eliminate the project's only dependency.

#### How should this be manually tested?

   1. Clone this repository
   2. Run `npm install`
   3. Run `npm run doc`
   4. Run `npm run serve:doc`
   5. Go to <http://localhost:3000/UserInterfaceTutorial.html>
   6. Interact with the file app (eg. add a file or create a new directory, add/delete a file, etc)

The files app should behave normally.

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

N/A

#### What are the relevant tickets?

N/A

#### TODOs / next steps:
